p2p/src/msg.rs

```diff
@@ -118,10 +118,19 @@ pub fn read_exact(
 	let sleep_time = time::Duration::from_micros(10);
 	let mut count = time::Duration::new(0, 0);
 
+	let exact = buf.len();
 	let mut read = 0;
 	loop {
 		match conn.read(buf) {
-			Ok(0) => break,
+			Ok(0) => {
+				if read < exact {
```

Comment:
Nice catch! What do you think about always return the error here? I don't see how we can get here with full (well, empty) buffer. We read and check the buffer below, if we are done we break https://github.com/mimblewimble/grin/blob/5a1d69744e19dafe023b58d1f8a15d3f37e1929b/p2p/src/msg.rs#L151. So we read again only if we expect to read at least 1 byte.

Reply:
I agree it looks like `if read < exact` will be always `true` and `break` will never be reached.

But that `break` make us 100% safe from an infinite loop, so, if it doesn't hurt, want to keep it there:)

btw, that comes from a habit to avoid static check tool complain, because normally I use static analysis tool for big projects. 

we could also need use static analysis tool for Grin.

Reply:
No, I mean we should unconditionally return error, so we exit loop here anyway, just now we have 2 branches which is harder to read and to reason about. 
reg static analysis - we use Rust already:) on a serious note I use `clippy` from time to time, but not sure it should be a part of CI.

Reply:
Sorry, don't understand. Do you mean just put a `return Err(io::Error::new(...))` there and remove `if read < exact {` and `break`?

Reply:
Right, I think we both think that if we read `0` here  it's an error for us